File: scripts/causal-cleanup.test.mjs
/** Loaded private driver/cleanup tests: all factories and clocks are fake. */
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import { syncBuiltinESMExports } from "node:module";
import os from "node:os";
import path from "node:path";
import { performance } from "node:perf_hooks";
import timers from "node:timers/promises";
import { fileURLToPath, pathToFileURL } from "node:url";
import { createGaps, finishBoth } from "./causal-cleanup-observer.mjs";
import { createObserver } from "./causal-cpu-observer.mjs";
import { driver, worker } from "./derive-causal-cleanup.mjs";

const material = process.argv[2];
assert.ok(material);
const original = fs.readFileSync(path.join(material, "position.mjs"), "utf8");
const bundle = fs.readFileSync(path.join(material, "worker.mjs"), "utf8");
const fixed = JSON.parse(original.match(/const originalRecipe = (.*);/)[1]);
const self = fileURLToPath(import.meta.url);
const derivedWorker = worker(bundle);
const match = derivedWorker.match(/cleanup: \(diagnostic\) => \{([\s\S]*?)\n {4}\}/);
assert.ok(match);
if (process.argv[3] === "--stub") {
	const [root, mode, orientation, fault] = process.argv.slice(4);
	const job = path.join(root, "job");
	fs.mkdirSync(job);
	let source = mode === "B" ? original : driver(original),
		body = match[1];
	if (fault === "getter-inside")
		source = source.replace(
			'if (row.group === "cold") run = makeArm(arm, row.mode);',
			'if (row.group === "cold") { performance2.now(); run = makeArm(arm, row.mode); }',
		);
	if (fault === "drop-cleanup")
		source = source.replace(
			"if (run !== shared) cleanupAll([run], sampleFailure, deep);",
			"if (false) cleanupAll([run], sampleFailure, deep);",
		);
	if (fault === "drop-release") body = body.replace("group.release();", "");
	if (fault === "swap-deep")
		body = body
			.replace("diagnostic.d1 =", "diagnostic.d9 =")
			.replace("diagnostic.d2 =", "diagnostic.d1 =")
			.replace("diagnostic.d9 =", "diagnostic.d2 =");
	if (fault === "missing-d4")
		body = body.replace("if (diagnostic) diagnostic.d4 = diagnostic.now(); // GAP", "");
	if (fault === "coordinate")
		source = source.replace(
			"gaps.begin({batch:batch2, arm, index,",
			"gaps.begin({batch:batch2, arm, index:index+1,",
		);
	if (fault === "swallow-primary")
		source = source.replace("sampleFailure = error;", "sampleFailure = undefined;");
	fs.writeFileSync(path.join(root, "position.mjs"), source);
	fs.writeFileSync(path.join(root, "observed.mjs"), source);
	fs.mkdirSync(path.join(root, "source/scripts"), { recursive: true });
	for (const name of ["causal-cpu-observer.mjs", "causal-cleanup-observer.mjs"])
		fs.copyFileSync(path.resolve("scripts", name), path.join(root, "source/scripts", name));
	const config = {
		row: { id: "cold-P2-summary", group: "cold", profile: "P2", mode: "summary" },
		kind: "control",
		control: true,
		orientation,
		output: job,
		scenarioPath: path.join(root, "P2-inputs.json"),
		mode: mode === "B" ? "B" : "Q",
		gapMode: fault === "mode-mismatch" ? "S" : mode,
		profile: "consumer",
		run: "fake",
		sourceDigest: "a".repeat(64),
	};
	let afterRecord = -1;
	let now = 0,
		cpu = 0,
		count = 0,
		release = 0,
		disconnects = 0,
		roots = 0,
		describes = 0,
		members = 0;
	const events = [],
		outputs = {},
		samples = [],
		loads = [],
		preflights = [];
	const cleanupFactory = new Function(
		"disconnect",
		"owner",
		"graph",
		`return (diagnostic)=>{${body}\n}`,
	);
	const modules = [0, 1].map((id) => ({
		RECIPE: structuredClone(fixed),
		preflight() {
			preflights.push(id);
			return { passed: true };
		},
		measurementArm(arm, m) {
			assert.equal(arm, "reference");
			assert.equal(m, "summary");
			count++;
			events.push("factory");
			const group = {
				add() {
					members++;
				},
				release() {
					release++;
					if (["dual", "swallow-primary", "cleanup"].includes(fault)) throw Error("cleanup");
				},
			};
			const cleanup = cleanupFactory(
				() => {
					disconnects++;
				},
				{
					roots: [
						{
							unsubscribe() {
								roots++;
							},
						},
					],
				},
				{
					topologyGroup() {
						return group;
					},
					describe() {
						describes++;
						return { nodes: [{ id: "one" }, { id: "two" }] };
					},
					find(id) {
						return id;
					},
				},
			);
			return { cleanup };
		},
	}));
	globalThis.cleanupStubModules = modules;
	for (const [i, name] of ["worker.mjs", "worker-copy.mjs"].entries()) {
		const stub = `const m=globalThis.cleanupStubModules[${i}];export const RECIPE=m.RECIPE;export const preflight=(...a)=>m.preflight(...a);export const measurementArm=(...a)=>m.measurementArm(...a);`;
		fs.writeFileSync(path.join(root, name), stub);
		fs.writeFileSync(path.join(root, "derived-" + name), stub);
	}
	const entry = spawnSync(
		"python3",
		[
			"-B",
			"-c",
			"import importlib.util,sys;from pathlib import Path;s=importlib.util.spec_from_file_location('c',sys.argv[1]);c=importlib.util.module_from_spec(s);s.loader.exec_module(c);print(c.entry(Path(sys.argv[2]),Path(sys.argv[3]),sys.argv[4]),end='')",
			path.resolve("scripts/causal-cleanup-diagnostic.py"),
			root,
			job,
			mode,
		],
		{ encoding: "utf8", timeout: 15000 },
	);
	assert.equal(entry.status, 0, entry.stderr);
	fs.writeFileSync(path.join(job, "entry.mjs"), entry.stdout);
	const realRead = fs.readFileSync,
		realWrite = fs.writeFileSync;
	fs.readFileSync = (p, ...args) =>
		p === path.join(job, "config.json")
			? JSON.stringify(config)
			: p === config.scenarioPath
				? "{}"
				: realRead(p, ...args);
	fs.writeFileSync = (p, s) => {
		if (fault === "write-failure" && p.endsWith("evidence.json")) throw Error("evidence-write");
		outputs[path.basename(p)] = JSON.parse(s);
	};
	fs.appendFileSync = (p, s) => {
		if (p.endsWith("entry-events.jsonl")) {
			loads.push(JSON.parse(s));
			return;
		}
		events.push("record");
		afterRecord = 0;
		if (["dual", "swallow-primary", "c0-dual", "d0-dual", "c1-dual"].includes(fault))
			throw Error("record");
		samples.push(JSON.parse(s));
	};
	timers.setImmediate = async () => {
		events.push("yield");
	};
	Object.defineProperty(performance, "now", {
		configurable: true,
		value: () => {
			events.push("clock");
			if (afterRecord >= 0) {
				const n = afterRecord++;
				if (
					(fault === "c0-dual" && n === 0) ||
					(fault === "d0-dual" && n === 1) ||
					(fault === "c1-dual" && n === 6)
				)
					throw Error("clock-failure");
			}
			return ++now;
		},
	});
	process.memoryUsage = () => {
		events.push("memory");
		return { heapUsed: 10, rss: 20 };
	};
	process.threadCpuUsage = () => {
		events.push("thread");
		return { user: ++cpu * 1000, system: cpu * 100 };
	};
	process.resourceUsage = () => {
		events.push("process");
		return {
			userCPUTime: cpu * 3000,
			systemCPUTime: cpu * 100,
			voluntaryContextSwitches: cpu,
			involuntaryContextSwitches: 0,
			minorPageFault: 0,
			majorPageFault: 0,
		};
	};
	syncBuiltinESMExports();
	let error = null;
	try {
		await import(pathToFileURL(path.join(job, "entry.mjs")).href);
	} catch (e) {
		error = e;
	}
	const messages = (e) => [String(e), ...(e?.errors ?? []).flatMap(messages)];
	if (
		[
			"dual",
			"cleanup",
			"write-failure",
			"swallow-primary",
			"c0-dual",
			"d0-dual",
			"c1-dual",
		].includes(fault)
	) {
		assert.ok(error);
		const msg = messages(error).join("\n");
		if (["dual", "swallow-primary"].includes(fault)) {
			assert.match(msg, /record/);
			assert.match(msg, /cleanup/);
		}
		if (["c0-dual", "d0-dual", "c1-dual"].includes(fault)) {
			assert.match(msg, /record/);
			assert.match(msg, /clock-failure/);
			assert.equal(release, 1);
		}
		if (fault === "write-failure") assert.match(msg, /evidence-write/);
		assert.ok(outputs["gaps.json"]);
		assert.equal(outputs["gaps.json"].complete, fault === "write-failure");
	} else {
		assert.equal(error, null, error?.stack);
		assert.equal(count, 2400);
		assert.equal(release, 2400);
		assert.equal(disconnects, 2400);
		assert.equal(roots, 2400);
		assert.equal(describes, 2400);
		assert.equal(members, 4800);
		assert.deepEqual(preflights, [0, 1]);
		assert.deepEqual(
			loads.map((x) => x.module),
			["n", "m0", "m1"],
		);
		assert.equal(samples.length, 2400);
		assert.ok(samples.every((s) => s.ms === 1));
		for (let i = 0; i < events.length; i++)
			if (events[i] === "factory")
				assert.deepEqual(events.slice(i - 2, i + 3), [
					"memory",
					"clock",
					"factory",
					"clock",
					"memory",
				]);
		if (mode === "B") {
			assert.equal(outputs["gaps.json"], undefined);
			assert.equal(outputs["evidence.json"], undefined);
		} else {
			const g = outputs["gaps.json"],
				e = outputs["evidence.json"];
			assert.equal(g.gapMode, mode);
			assert.equal(g.samples.length, 2400);
			assert.equal(g.complete, true);
			assert.equal(e.windows.length, 12);
			for (let i = 0; i < 2400; i++) {
				const r = g.samples[i],
					s = samples[i];
				for (const k of ["batch", "arm", "index", "phase", "start", "end"])
					assert.equal(r[k], s[k]);
				const ts = ["y0", "y1", "m0", "start", "end", "m1", "r0", "r1", "c0", "c1"].map(
					(k) => r[k],
				);
				assert.ok(ts.every(Number.isFinite));
				assert.deepEqual(
					ts,
					[...ts].sort((a, b) => a - b),
				);
				if (mode === "D") {
					assert.deepEqual(Object.keys(r.deep), ["d0", "d1", "d2", "d3", "d4"]);
					const d = Object.values(r.deep);
					assert.deepEqual(
						d,
						[...d].sort((a, b) => a - b),
					);
					assert.ok(r.c0 < d[0] && d[4] < r.c1);
				} else assert.equal(r.deep, undefined);
			}
			assert.equal(events.filter((x) => x === "thread").length, 24);
			assert.equal(events.filter((x) => x === "process").length, 24);
			assert.equal(
				events.filter((x) => x === "clock").length,
				1 + 2400 * (2 + 8 + (mode === "D" ? 5 : 0)) + 48,
			);
			realWrite(path.join(root, "synthetic.json"), JSON.stringify({ g, samples, e, mode }));
		}
	}
	console.log("CLEANUP_LOADED_OK", mode, orientation, fault);
} else {
	const root = fs.mkdtempSync(path.join(os.tmpdir(), "cleanup-offline-"));
	let positives = 0,
		mutants = 0;
	try {
		for (const mode of ["B", "S", "D"])
			for (const orientation of ["U", "V"]) {
				const dir = path.join(root, mode + orientation);
				fs.mkdirSync(dir);
				const r = spawnSync(
					process.execPath,
					[self, material, "--stub", dir, mode, orientation, "none"],
					{ encoding: "utf8", timeout: 20000 },
				);
				assert.equal(r.status, 0, r.stdout + r.stderr);
				positives++;
				if (mode !== "B") {
					const v = spawnSync(
						"python3",
						[
							"-B",
							"scripts/causal-cleanup.test.py",
							"--synthetic",
							path.join(dir, "synthetic.json"),
						],
						{ encoding: "utf8", timeout: 20000 },
					);
					assert.equal(v.status, 0, v.stdout + v.stderr);
				}
			}
		for (const fault of [
			"getter-inside",
			"drop-cleanup",
			"drop-release",
			"swap-deep",
			"missing-d4",
			"coordinate",
			"mode-mismatch",
			"swallow-primary",
			"dual",
			"cleanup",
			"write-failure",
			"c0-dual",
			"d0-dual",
			"c1-dual",
		]) {
			const dir = path.join(root, fault);
			fs.mkdirSync(dir);
			const r = spawnSync(process.execPath, [self, material, "--stub", dir, "D", "U", fault], {
				encoding: "utf8",
				timeout: 20000,
			});
			if (["dual", "cleanup", "write-failure", "c0-dual", "d0-dual", "c1-dual"].includes(fault)) {
				assert.equal(r.status, 0, r.stdout + r.stderr);
				positives++;
			} else {
				assert.notEqual(r.status, 0, fault + " survived");
				mutants++;
			}
		}
		assert.throws(() => driver(original + " "));
		assert.throws(() => worker(bundle.replace("group.release();", "group.noRelease();")));
		const errors = [];
		assert.throws(
			() =>
				finishBoth(
					{
						finish() {
							errors.push("cpu");
							throw Error("cpu-write");
						},
					},
					{
						finish() {
							errors.push("gap");
							throw Error("gap-write");
						},
					},
					Error("primary"),
				),
			(e) => e.errors.length === 3,
		);
		assert.deepEqual(errors, ["cpu", "gap"]);
		console.log(
			JSON.stringify({
				passed: true,
				loadedPositives: positives,
				loadedMutants: mutants,
				newConsumerSamples: 0,
			}),
		);
	} finally {
		fs.rmSync(root, { recursive: true, force: true });
	}
}

File: scripts/causal-cleanup.test.py
"""Offline independent arithmetic/source qualifications; no consumer execution."""
import sys,json,importlib.util,copy,tempfile,subprocess
from pathlib import Path
spec=importlib.util.spec_from_file_location('v',Path(__file__).with_name('verify-causal-cleanup.py'));v=importlib.util.module_from_spec(spec);spec.loader.exec_module(v)
def rejected(call):
    try:call()
    except (AssertionError,KeyError,ValueError,TypeError):return
    raise AssertionError('negative survived')
if sys.argv[1]=='--synthetic':
    a=json.loads(Path(sys.argv[2]).read_text());g,s,e,mode=[a[k] for k in ('g','samples','e','mode')]
    out=v.verify_gaps(g,s,e,mode);v.cpu.verify_cpu(e,s)
    assert len(out)==12 and out[0]['segments']['yield']==dict(sumMs=100,p50=1,p95=1)
    if mode=='D':assert out[0]['segments']['group-release']==dict(sumMs=100,p50=1,p95=1)
    negatives=0
    for mutate in [lambda x:x.update(units='seconds'),lambda x:x.update(gapMode='B'),lambda x:x.update(complete=False),lambda x:x.update(active={}),lambda x:x['samples'].pop(),lambda x:x['samples'].append(x['samples'][0]),lambda x:x['samples'][0].update(index=1),lambda x:x['samples'][0].update(batch=True),lambda x:x['samples'][0].update(y0=-1),lambda x:x['samples'][0].update(y0=float('nan')),lambda x:x['samples'][0].update(y0=x['samples'][0]['y1']+1),lambda x:x['samples'][0].update(c1=1e99),lambda x:x['samples'][0].update(start=0),lambda x:x['samples'][0].update(extra=1)]:
        x=copy.deepcopy(g);mutate(x);rejected(lambda:v.verify_gaps(x,s,e,mode));negatives+=1
    if mode=='D':
        for mutate in [lambda x:x['samples'][0]['deep'].pop('d4'),lambda x:x['samples'][0]['deep'].update(d1=-1),lambda x:x['samples'][0]['deep'].update(d4=1e99),lambda x:x['samples'][0]['deep'].update(d0=float('inf')),lambda x:x['samples'][0]['deep'].update(d3=True)]:
            x=copy.deepcopy(g);mutate(x);rejected(lambda:v.verify_gaps(x,s,e,mode));negatives+=1
    if mode=='D' and e['orientation']=='U':
        import shutil
        with tempfile.TemporaryDirectory() as temp:
            root=Path(temp).resolve();physical=root
            material=Path('archive/evals/causal-cpu-wall-diagnostic-v1/run')
            for name in ['worker.mjs','worker-copy.mjs','position.mjs','P2-inputs.json','reference-qualification.json']:shutil.copyfile(material/name,root/name)
            for kind,source,target in [('driver','position.mjs','observed.mjs'),('worker','worker.mjs','derived-worker.mjs'),('worker','worker-copy.mjs','derived-worker-copy.mjs')]:subprocess.run(['node','scripts/derive-causal-cleanup.mjs',kind,str(root/source),str(root/target)],check=True)
            put=lambda p,x:p.write_text(json.dumps(x)+'\n')
            for name in v.SOURCES:
                target=root/'source/scripts'/name;target.parent.mkdir(parents=True,exist_ok=True);shutil.copyfile(Path('scripts')/name,target)
            approval=json.loads(Path('docs/design/causal-cleanup-segments-v1-implementation/approval.json').read_text());put(root/'approval.json',approval);shutil.copyfile(approval['design'],root/'approved-design.md')
            hashes={'scripts/'+n:v.sha(root/'source/scripts'/n) for n in v.SOURCES}
            (root/'qualification').mkdir();put(root/'qualification/qualification.json',dict(passed=True,sourceDigests=hashes,designDigest=approval['designDigest'],logs={}))
            reservation=dict(status='reserved',executionRoot=str(root),schedule=v.SCHEDULE,implicitNodeOptions={},runtime=dict(node='v24.18.0',v8='13.6.233.17-node.50',platform='darwin',arch='arm64'),approvalDigest=v.sha(root/'approval.json'),qualificationDigest=v.sha(root/'qualification/qualification.json'),sourceDigests=hashes,startMonotonic=100,run='fake-run',executable='/fake/node',wakeAtStart='fake-wake')
            put(root/'reservation.json',reservation)
            put(root/'materials.json',{n:v.sha(root/n) for n in ['worker.mjs','worker-copy.mjs','derived-worker.mjs','derived-worker-copy.mjs','position.mjs','observed.mjs','P2-inputs.json','reference-qualification.json']})
            for position,cell in enumerate(v.SCHEDULE):
                m,o=cell.split('-');d=root/f'{position:02d}-{cell}';d.mkdir();pid=1000+position
                (d/'entry.mjs').write_text(v.expected_entry(root,d,m));digest=v.sha(d/'entry.mjs')
                config=dict(output=str(d),run=f'fake-run-{position}',sourceDigest=digest,mode='B' if m=='B' else 'Q',gapMode=m,profile='consumer',orientation=o,row=v.old.ROW,kind='control',control=True,scenarioPath=str(root/'P2-inputs.json'));put(d/'config.json',config)
                dispatch=dict(argv=['/fake/node',*v.FLAGS,str(d/'entry.mjs')],cwd=str(d),position=position,cell=cell,wakeBefore='fake-wake',startMonotonic=101+position,startWall=1001+position)
                put(d/'dispatch.json',dispatch);put(d/'exit.json',dict(**dispatch,pid=pid,exitCode=0,childError=None,wakeError=None,finalizationErrors=[],wakeAfter='fake-wake',endMonotonic=101.1+position,endWall=1001.1+position))
                ss=copy.deepcopy(s);gg=copy.deepcopy(g);ee=copy.deepcopy(e)
                for i,r in enumerate(ss):
                    block,index=divmod(i,400);batch,slot=divmod(block,2);arm=v.old.order(o,batch)[slot]
                    r['arm']=arm;r['segment']=f'cold-P2-summary/{batch}/{arm}/{index}';gg['samples'][i]['arm']=arm
                    if m=='S':gg['samples'][i].pop('deep')
                (d/'samples.jsonl').write_text(''.join(json.dumps(r)+'\n' for r in ss));put(d/'completion.json',dict(completed=True,samples=2400));(d/'stdout.log').write_text('PRESET_PERFORMANCE_ROW_DONE cold-P2-summary 2400');(d/'stderr.log').write_text('')
                (d/'entry-events.jsonl').write_text(''.join(json.dumps(dict(module=n,pid=pid))+'\n' for n in ['n','m0','m1']))
                put(d/'preflight-first.json',dict(module=0,pid=pid));put(d/'preflight-order.json',dict(modules=[0,1],pid=pid));put(d/'preflight.json',dict(passed=True))
                put(d/'worker.json',dict(pid=pid,node='v24.18.0',control=True,row=v.old.ROW,recipe={**v.old.RECIPE,'orders':[v.old.order(o,b) for b in range(3)]}))
                if m!='B':
                    ee['identity']=dict(run=config['run'],pid=pid,node='v24.18.0',v8='13.6.233.17-node.50',sourceDigest=digest);ee['orientation']=o
                    gg['identity']=dict(run=config['run'],pid=pid,sourceDigest=digest);gg['gapMode']=m
                    put(d/'evidence.json',ee);put(d/'gaps.json',gg)
                row=v.verify_job(root,position);put(d/'independent-command.json',dict(exitCode=0,stdout=json.dumps(dict(evidenceValid=True,**row)),stderr=''))
            result=dict(attempted=list(range(12)),notRun=[],retries=0,performanceQualification=False,fixture=None,status='complete-diagnostic',stopReason=None,rows=[dict(position=i,cell=cell,samples=2400) for i,cell in enumerate(v.SCHEDULE)],samples=28800,capturedSamples=28800,elapsedSeconds=20)
            put(root/'result.json',result)
            def index():put(root/'artifact-index.json',dict(files={str(p.relative_to(root)):v.sha(p) for p in root.rglob('*') if p.is_file() and p.name!='artifact-index.json'}))
            index();assert v.verify(root)['samples']==28800
            d=root/'02-D-U';config=json.loads((d/'config.json').read_text());bad={**config,'run':'wrong'};put(d/'config.json',bad);rejected(lambda:v.verify_job(root,2));put(d/'config.json',config)
            gaps=json.loads((d/'gaps.json').read_text());bad=copy.deepcopy(gaps);bad['identity']['pid']=999;put(d/'gaps.json',bad);rejected(lambda:v.verify_job(root,2));put(d/'gaps.json',gaps)
            for name in ['hidden/exit.json','source/deeper/gaps.json']:
                p=root/name;p.parent.mkdir(parents=True,exist_ok=True);put(p,{});index();rejected(lambda:v.verify(root));p.unlink()
                if p.parent.name=='hidden':p.parent.rmdir()
            # A failed last row still has exact entry/dispatch identity and must reproduce its failed validation.
            d=root/'11-B-U';completion=json.loads((d/'completion.json').read_text());put(d/'completion.json',dict(completed=False,samples=2400))
            result.update(status='incomplete',stopReason='failed validation',rows=result['rows'][:-1],samples=26400);put(root/'result.json',result)
            failure=dict(stage='validate',position=11,error='failed validation',verifiedRows=11);put(root/'failure.json',failure);index();assert v.verify(root)['failedJob']['validationFailureReproduced'] is True
            failure['position']=10;put(root/'failure.json',failure);index();rejected(lambda:v.verify(root));failure['position']=11;put(root/'failure.json',failure)
            config=json.loads((d/'config.json').read_text());config['run']='forged-failed-run';put(d/'config.json',config);index();rejected(lambda:v.verify(root))
            print(json.dumps(dict(syntheticWholeBatchVerified=True,syntheticFailedBatchVerified=True,wholeJobNegatives=6,realConsumerSamples=0)))
    print(json.dumps(dict(passed=True,negatives=negatives)))
else:
    material=Path(sys.argv[1]);original=(material/'position.mjs').read_text();worker=(material/'worker.mjs').read_text()
    with tempfile.TemporaryDirectory() as tmp:
        tmp=Path(tmp)
        for kind,name,target in [('driver','position.mjs','d.mjs'),('worker','worker.mjs','w.mjs')]:subprocess.run(['node','scripts/derive-causal-cleanup.mjs',kind,str(material/name),str(tmp/target)],check=True)
        d=(tmp/'d.mjs').read_text();w=(tmp/'w.mjs').read_text();v.check_source(original,d,worker,w)
        for a,b in [('const start = performance2.now();','const start = 0;'),('gap.y0 = performance2.now();','gap.y0 = 0;'),('gap.r0 = performance2.now();','gap.r0 = performance2.now(); gap.extra = 1;'),('gaps.append(gap, sampleFailure);','gaps.append({...gap}, sampleFailure);'),('run?.cleanup(diagnostic);','run?.cleanup();')]:rejected(lambda:v.check_source(original,d.replace(a,b),worker,w))
        rejected(lambda:v.check_source(original,d,worker,w.replace('group.release();','')))
    # Independent expected entry must match generator for every exact cell, without dispatch.
    spec=importlib.util.spec_from_file_location('collector',Path(__file__).with_name('causal-cleanup-diagnostic.py'));c=importlib.util.module_from_spec(spec);spec.loader.exec_module(c)
    for mode in ['B','S','D']:assert c.entry(Path('/frozen/run'),Path('/frozen/run/00-job'),mode)==v.expected_entry(Path('/frozen/run'),Path('/frozen/run/00-job'),mode)
    assert c.SCHEDULE==v.SCHEDULE and c.SOURCES==['scripts/'+n for n in v.SOURCES]
    # Preparation refusal must consume only the reservation, never spawn a consumer.
    import shutil
    approval=Path('docs/design/causal-cleanup-segments-v1-implementation/approval.json')
    for field,value in [('designDigest','sha256:bad'),('maxOutputBytes',1),('checkIntervalMs',1),('maxFixtureProcesses',1)]:
        with tempfile.TemporaryDirectory() as directory:
            root=Path(directory).resolve();doc=root/'doc';doc.mkdir();out=root/'run'
            a=json.loads(approval.read_text());a[field]=value
            target=root/a['design'];target.parent.mkdir(parents=True);shutil.copyfile(a['design'],target)
            c.put(doc/'approval.json',a);c.ROOT=root;c.DOC=doc;c.OUTPUT=out
            assert c.run()==1 and c.read(out/'result.json')['attempted']==[]
            assert v.verify(out)['processes']==0
            def reindex():c.put(out/'artifact-index.json',dict(files={str(p.relative_to(out)):v.sha(p) for p in out.rglob('*') if p.is_file() and p.name!='artifact-index.json'}))
            result=c.read(out/'result.json');result['capturedSamples']=1;c.put(out/'result.json',result);reindex();rejected(lambda:v.verify(out));result['capturedSamples']=0;c.put(out/'result.json',result)
            hidden=out/'hidden/deep/exit.json';hidden.parent.mkdir(parents=True);c.put(hidden,{});reindex();rejected(lambda:v.verify(out))
    print(json.dumps(dict(passed=True,sourceNegatives=6,entryModes=3,preparationRefusals=4,inventoryNegatives=8,newConsumerSamples=0)))
